File: lab11Calificado/Lab11_PLANTILLA/TestSoftSolution/TestSoftPersistance/src/main/java/pe/edu/pucp/testsoft/persistance/alumno/Impl/AlumnoImpl.java
package pe.edu.pucp.testsoft.persistance.alumno.Impl;

import pe.edu.pucp.testsoft.config.DBManager;
import pe.edu.pucp.testsoft.model.Alumno;
import pe.edu.pucp.testsoft.persistance.alumno.DAO.AlumnoDAO;

import java.sql.ResultSet;
import java.sql.Types;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// DAO de Alumno. TODO acceso a datos pasa por Stored Procedures (nunca SQL inline).
// Parámetros POSICIONALES: la clave del Map es la posición del ? en {call sp(?,?,...)}.
public class AlumnoImpl implements AlumnoDAO {

    @Override
    public int insertar(Alumno alumno) {
        Map<Integer, Object> entrada = new HashMap<>();
        Map<Integer, Object> salida  = new HashMap<>();
        // SP_INSERTAR_ALUMNO(OUT p_id, IN p_codigo, p_nombre, p_apellidos, p_correo, p_estado)
        salida.put(1, Types.INTEGER);              // OUT va PRIMERO (posición 1)
        entrada.put(2, alumno.getCodigo());
        entrada.put(3, alumno.getNombre());
        entrada.put(4, alumno.getApellidos());
        entrada.put(5, alumno.getCorreo());
        entrada.put(6, alumno.getEstado());
        DBManager.getInstance().ejecutarProcedimiento("SP_INSERTAR_ALUMNO", entrada, salida);
        alumno.setId((int) salida.get(1));         // recuperar el id auto_increment
        return alumno.getId();
    }

    @Override
    public int modificar(Alumno alumno) {
        Map<Integer, Object> entrada = new HashMap<>();
        // SP_MODIFICAR_ALUMNO(IN p_id, p_codigo, p_nombre, p_apellidos, p_correo, p_estado)
        entrada.put(1, alumno.getId());
        entrada.put(2, alumno.getCodigo());
        entrada.put(3, alumno.getNombre());
        entrada.put(4, alumno.getApellidos());
        entrada.put(5, alumno.getCorreo());
        entrada.put(6, alumno.getEstado());
        return DBManager.getInstance().ejecutarProcedimiento("SP_MODIFICAR_ALUMNO", entrada, null);
    }

    @Override
    public int eliminar(int id) {
        Map<Integer, Object> entrada = new HashMap<>();
        entrada.put(1, id);                        // baja lógica (estado='I') dentro del SP
        return DBManager.getInstance().ejecutarProcedimiento("SP_ELIMINAR_ALUMNO", entrada, null);
    }

    @Override
    public Alumno buscarPorId(int id) {
        Alumno alumno = null;
        Map<Integer, Object> entrada = new HashMap<>();
        entrada.put(1, id);
        try (DBManager.ResultadoConsulta rc =
                     DBManager.getInstance().ejecutarProcedimientoLectura("SP_OBTENER_ALUMNO_POR_ID", entrada)) {
            ResultSet rs = rc.getRs();
            if (rs.next()) alumno = mapear(rs);     // 0 o 1 fila -> if, no while
        } catch (Exception ex) {
            System.out.println("Error al buscar alumno por id: " + ex.getMessage());
        }
        return alumno;
    }

    @Override
    public List<Alumno> listarTodos() {
        return buscarPorNombreApellido("");        // reusa el SP de búsqueda con texto vacío
    }

    @Override
    public List<Alumno> buscarPorNombreApellido(String texto) {
        List<Alumno> lista = new ArrayList<>();
        Map<Integer, Object> entrada = new HashMap<>();
        entrada.put(1, texto);
        try (DBManager.ResultadoConsulta rc =
                     DBManager.getInstance().ejecutarProcedimientoLectura("SP_LISTAR_ALUMNOS_X_NOMBRE_APELLIDO", entrada)) {
            ResultSet rs = rc.getRs();
            while (rs.next()) lista.add(mapear(rs));
        } catch (Exception ex) {
            System.out.println("Error al listar alumnos: " + ex.getMessage());
        }
        return lista;
    }

    @Override
    public int buscarPorCodigo(String codigo) {
        // SELECT INTO: el SP mete el id en un OUT en vez de devolver filas -> NO usar lectura.
        Map<Integer, Object> entrada = new HashMap<>();
        Map<Integer, Object> salida  = new HashMap<>();
        entrada.put(1, codigo);
        salida.put(2, Types.INTEGER);
        DBManager.getInstance().ejecutarProcedimiento("SP_BUSCAR_ALUMNO_POR_CODIGO", entrada, salida);
        Object valor = salida.get(2);
        if (valor == null) return 0;               // no encontrado
        return (int) valor;
    }

    // Mapeo fila -> objeto, en un solo lugar para no repetir.
    private Alumno mapear(ResultSet rs) throws Exception {
        Alumno a = new Alumno();
        a.setId(rs.getInt("id"));
        a.setCodigo(rs.getString("codigo"));
        a.setNombre(rs.getString("nombre"));
        a.setApellidos(rs.getString("apellidos"));
        a.setCorreo(rs.getString("correo"));
        a.setEstado(rs.getString("estado"));       // CHAR(1) -> getString da "A"/"I"
        return a;
    }
}

File: lab11Calificado/Lab11_PLANTILLA/ExamenParcialCSharp/GestionAlumnosPersistance/Alumnos/Impl/AlumnoImpl.cs
// =====================================================================
// AlumnoImpl.cs — DAO de Alumno
// =====================================================================
// Este archivo muestra TODOS los patrones básicos de acceso a datos:
// 1. INSERTAR con parámetro OUT (recuperar ID auto_increment)
// 2. MODIFICAR y ELIMINAR (operaciones simples)
// 3. BUSCAR POR ID con DataReader (EjecutarProcedimientoLectura)
// 4. LISTAR TODOS con DataReader
// 5. SELECT INTO — buscar sin DataReader, usando parámetro OUT
// =====================================================================

using GestionAlumnosDBManager;
using GestionAlumnosModel.Alumno;
using GestionAlumnosPersistance.Alumnos.DAO;
using System.Data;
using System.Data.Common;

namespace GestionAlumnosPersistance.Alumnos.Impl
{
    public class AlumnoImpl : AlumnoDAO
    {
        public int Insertar(Alumno alumno)
        {
            DbParameter[] parametros = new DbParameter[6];

            // Los nombres DEBEN coincidir con los del SP en MySQL
            parametros[1] = DBManager.Instance.CreateParam("p_codigo", DbType.String, alumno.Codigo, ParameterDirection.Input);
            parametros[2] = DBManager.Instance.CreateParam("p_nombre", DbType.String, alumno.Nombre, ParameterDirection.Input);
            parametros[3] = DBManager.Instance.CreateParam("p_apellidos", DbType.String, alumno.Apellidos, ParameterDirection.Input);
            parametros[4] = DBManager.Instance.CreateParam("p_correo", DbType.String, alumno.Correo, ParameterDirection.Input);
            parametros[5] = DBManager.Instance.CreateParam("p_estado", DbType.String, alumno.Estado, ParameterDirection.Input);

            // Parámetro OUTPUT: valor = null, dirección = Output
            // Después de ejecutar el SP, este parámetro contendrá el ID generado
            parametros[0] = DBManager.Instance.CreateParam("p_id", DbType.Int32, null, ParameterDirection.Output);

            // EjecutarProcedimiento → para INSERT/UPDATE/DELETE (no retorna DataReader)
            DBManager.Instance.EjecutarProcedimiento("SP_INSERTAR_ALUMNO", parametros);

            // Leer el valor OUT: el ID que la BD generó con auto_increment
            alumno.Id = Convert.ToInt32(parametros[0].Value);
            return alumno.Id;
        }

        public int Modificar(Alumno alumno)
        {
            DbParameter[] parametros = new DbParameter[6];
            parametros[0] = DBManager.Instance.CreateParam("p_id", DbType.Int32, alumno.Id, ParameterDirection.Input);
            parametros[1] = DBManager.Instance.CreateParam("p_codigo", DbType.String, alumno.Codigo, ParameterDirection.Input);
            parametros[2] = DBManager.Instance.CreateParam("p_nombre", DbType.String, alumno.Nombre, ParameterDirection.Input);
            parametros[3] = DBManager.Instance.CreateParam("p_apellidos", DbType.String, alumno.Apellidos, ParameterDirection.Input);
            parametros[4] = DBManager.Instance.CreateParam("p_correo", DbType.String, alumno.Correo, ParameterDirection.Input);
            parametros[5] = DBManager.Instance.CreateParam("p_estado", DbType.String, alumno.Estado, ParameterDirection.Input);

            return DBManager.Instance.EjecutarProcedimiento("SP_MODIFICAR_ALUMNO", parametros);
        }

        public int Eliminar(int id)
        {
            DbParameter[] parametros = new DbParameter[1];
            parametros[0] = DBManager.Instance.CreateParam("p_id", DbType.Int32, id, ParameterDirection.Input);
            return DBManager.Instance.EjecutarProcedimiento("eliminar_alumno", parametros);
        }


        public Alumno BuscarPorId(int id)
        {
            Alumno? alumno = null; // '?' = nullable, puede ser null si no se encuentra

            DbParameter[] parametros = new DbParameter[1];
            parametros[0] = DBManager.Instance.CreateParam("p_id", DbType.Int32, id, ParameterDirection.Input);

            // 'using' cierra el DataReader al salir del bloque
            using DbDataReader lector = DBManager.Instance.EjecutarProcedimientoLectura("SP_OBTENER_ALUMNO_POR_ID", parametros);

            // Para BuscarPorId, esperamos 0 o 1 fila, por eso usamos 'if' en vez de 'while'.
            if (lector.Read())
            {
                alumno = new Alumno();
                alumno.Id = lector.GetInt32(lector.GetOrdinal("id"));
                alumno.Codigo = lector.GetString(lector.GetOrdinal("codigo"));
                alumno.Nombre = lector.GetString(lector.GetOrdinal("nombre"));
                alumno.Apellidos = lector.GetString(lector.GetOrdinal("apellidos"));
                alumno.Correo = lector.GetString(lector.GetOrdinal("correo"));
                alumno.Estado = lector.GetChar(lector.GetOrdinal("estado"));
                return alumno;
            }
            return null;
        }

        public List<Alumno> ListarTodos()
        {
            List<Alumno> lista = new List<Alumno>();
            
            DbParameter[] parametros = new DbParameter[1];
            // Parámetro 1: IN — el nombre a buscar, en este caso vacio para listar todos
            String artificio = ""; 
            parametros[0] = DBManager.Instance.CreateParam("p_texto", DbType.String, artificio, ParameterDirection.Input);
            // null como parámetros → el SP no recibe parámetros
            using DbDataReader lector = DBManager.Instance.EjecutarProcedimientoLectura("SP_LISTAR_ALUMNOS_X_NOMBRE_APELLIDO", parametros);

            // while → recorre TODAS las filas
            while (lector.Read())
            {
                Alumno alumno = new Alumno();
                alumno.Id = lector.GetInt32(lector.GetOrdinal("id"));
                alumno.Codigo = lector.GetString(lector.GetOrdinal("codigo"));
                alumno.Nombre = lector.GetString(lector.GetOrdinal("nombre"));
                alumno.Apellidos = lector.GetString(lector.GetOrdinal("apellidos"));
                alumno.Correo = lector.GetString(lector.GetOrdinal("correo"));
                alumno.Estado= lector.GetChar(lector.GetOrdinal("estado"));

                lista.Add(alumno);
            }
            return lista;
        }


        public int BuscarPorCodigo(string codigo)
        {
            DbParameter[] parametros = new DbParameter[2];

            // Parámetro 1: IN — el código a buscar
            parametros[0] = DBManager.Instance.CreateParam("p_codigo", DbType.String, codigo, ParameterDirection.Input);

            // Parámetro 2: OUT — aquí el SP devolverá el ID encontrado
            parametros[1] = DBManager.Instance.CreateParam("p_id", DbType.Int32, null, ParameterDirection.Output);

            // EjecutarProcedimiento (SIN Lectura) — porque SELECT INTO no retorna filas
            DBManager.Instance.EjecutarProcedimiento("buscar_alumno_por_codigo", parametros);

            // Verificar si encontró algo:
            // - DBNull.Value → el SELECT INTO no encontró ninguna fila (el SET p_id = NULL se mantiene)
            // - null → el parámetro no fue asignado
            if (parametros[1].Value == null || parametros[1].Value == DBNull.Value)
                return 0; // No encontrado

            return Convert.ToInt32(parametros[1].Value);
        }

        public List<Alumno> BuscarPorNombreApellido(String texto)
        {
            List<Alumno> alumnos = new List<Alumno>();

            DbParameter[] parametros = new DbParameter[1];
            parametros[0] = DBManager.Instance.CreateParam("p_texto", DbType.String, texto, ParameterDirection.Input);

            // 'using' cierra el DataReader al salir del bloque
            using DbDataReader lector = DBManager.Instance.EjecutarProcedimientoLectura("SP_LISTAR_ALUMNOS_X_NOMBRE_APELLIDO", parametros);

            // lector.Read() → avanza a la siguiente fila. Retorna false si no hay más.
            while (lector.Read())
            {
                Alumno alumno = new Alumno();
                alumno = new Alumno();
                alumno.Id = lector.GetInt32(lector.GetOrdinal("id"));
                alumno.Codigo = lector.GetString(lector.GetOrdinal("codigo"));
                alumno.Nombre = lector.GetString(lector.GetOrdinal("nombre"));
                alumno.Apellidos = lector.GetString(lector.GetOrdinal("apellidos"));
                alumno.Correo = lector.GetString(lector.GetOrdinal("correo"));
                alumno.Estado = lector.GetChar(lector.GetOrdinal("estado"));
                alumnos.Add(alumno);
            }
            return alumnos;
        }
    }
}
